### PR TITLE
Implement enough stuctures to lower a comparison with a constant int.

### PR DESCRIPTION
**Pull request by vext01**
TLDR: Implement enough stuff to lower an instruction like:
```
$3: t0 = ge($2: t0, I32(2000))
```
In doing so, I've noticed a few oddities with our intial design which
I've also addressed:

 - It seems odd to have an Operand as a Rvalue. It makes more sense to
   lift Operand's constituent bits directly into the Rvalue enum. In
   doing so we can reduce the number of "this can't happen"s later. E.g.
   we know something can only accept a constant, but we'd have to use an
   Operand and consider the Local case too.

   Operand still remains for operations which can accept both a constant
   *or* a local.

 - Instead of having binary integer operations as top-level rvalues, group them
   under a BinaryOp enum. This will allow us to more easily treat binary
   ops uniformly based on structure.

   Nullary and Unary ops will follow at some point.

 - Similarly for constant integers. Group them into a ConstantInt enum.

 - TIR operations are polymorphic, so the Add operation should Display
   as "add", not "int_add".

Additionally, I've made the Display for an unimplemented statement
lowering include the stringified MIR of the statement, so I can easily
see what lowering need to be implemented next.

Companion compiler PR will come soon.

### CODE REVIEW

File: ykpack/src/types.rs [hunk 1]
```diff
@@ -225,21 +227,85 @@ impl Display for Operand {
...
 }
 
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug, Clone)]
+pub enum ConstantInt {
+    UnsignedInt(UnsignedInt),
+    SignedInt(SignedInt),
+}
+
+impl ConstantInt {
+    // When constructing constant integers, truncation is deliberate.
```
Comment (ltratt) [inline ykpack/src/types.rs]: What does this mean?

Comment (ltratt) [reply ykpack/src/types.rs]: You might be able to do something like:

```rust
impl From<uwhatever> for ConstantInt {
    fn from(x: uwhatever) -> Self { ... }
}
```

and then avoid having to call differently named functions.

Comment (vext01) [reply ykpack/src/types.rs]: > What does this mean?

It's justification for the excessive use of `as`.

> You might be able to do something like:

Can I special-case u128? It breaks the mould sadly.

Comment (vext01) [reply ykpack/src/types.rs]: > Can I special-case u128? It breaks the mould sadly.

I don't think it's going to work. Each different type lowers to a different ykpack variant, so each type needs different functionality anyway.

Comment (ltratt) [reply ykpack/src/types.rs]: >Each different type lowers to a different ykpack variant, so each type needs different functionality anyway.

OK.

Can we make the comment a bit less gnomic?

Comment (vext01) [reply ykpack/src/types.rs]: Let me try some macros.

**Comment by vext01** [comment]
The last commit is the least offensive solution I could come up with.

Any good?

**Comment by ltratt** [comment]
Do we want to test all the integer conversions instead of just 1? And do we want to test the `128` case in particular?

**Comment by vext01** [comment]
> Do we want to test all the integer conversions instead of just 1?

I think *all* of them might be excessive, but I think you are right that testing a 128 case might be good.

**Comment by ltratt** [comment]
OK.

**Comment by vext01** [comment]
Last push tests a few more cases, including -ve values where incorrect truncation could yield a +ve value.

File: ykpack/src/types.rs [hunk 2]
```diff
@@ -524,9 +524,32 @@ mod tests {
...
+
+    #[test]
+    fn const_u64_from_bits() {
+        let v = std::u64::MAX;
+        let cst = ConstantInt::u64_from_bits(v as u128);
+        assert_eq!(cst, ConstantInt::UnsignedInt(UnsignedInt::U64(v)));
+    }
+
+    #[test]
+    fn const_i32_from_bits() {
```
Comment (ltratt) [inline ykpack/src/types.rs]: Let's order these low to high, so swap the 32 and 64 bit tests around?

**Comment by vext01** [comment]
> Let's order these low to high, so swap the 32 and 64 bit tests around?

Fixed

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.

**Comment by vext01** [comment]
Sorry about that. Shall I squash?

**Comment by ltratt** [comment]
Please squash.

**Comment by vext01** [comment]
splat.
